Make the metric in the buckets_path parameter optional

salvatore-campagna opened the pull request:
With this change the metric field name becomes optional if the
`bukets_path` is pointing to a multi-value aggregation with a single
metric field. Normally the full path would be required including
the aggregation name followed by the metric field.

If the metric is no specified in the path and the multi-value
aggregation computes more than one value an error is thrown.

The old notation is still supported for backward compatibility in case
the full path is specified and the target multi-value aggregation
computes a single value.

Resolves #72983.

salvatore-campagna committed:
Make the metric in the buckets_path parameter optional

With this change the metric field name becomes optional if the
bukets_path is pointing to a multi-value aggregation with a single
metric field. Normally the full path would be required including
the aggregation name followed by the metric field.

If the metric is no specified in the path and the multi-value
aggregation computes more than one value an error is thrown.

The old notation is still supported for backward compatibility in case
the full path is specified and the target multi-value aggreagation
computes a single value.

elasticmachine commented:
Pinging @elastic/es-analytics-geo (Team:Analytics)

salvatore-campagna committed:
Update docs/changelog/87220.yaml

elasticsearchmachine commented:
Hi @salvatore-campagna, I've created a changelog YAML for you.

salvatore-campagna committed:
fix: code format violations

salvatore-campagna committed:
fix: metric full path is optional therefore we do not throw

salvatore-campagna committed:
fix: limit version from 8.4.0 on

salvatore-campagna commented:
@not-napoleon According to my understanding those aggregations are the only ones where we can apply this kind of "shortcut" for the path. That is because they return an array as a result and specifying a metric might not be necessary. For those which return an object (like for instance extended stats) we still need the full path.

elasticmachine commented:
Pinging @elastic/clients-team (Team:Clients)

not-napoleon reviewed (changes requested):
The yaml tests are good, but I'd like to see some java tests as well.  As noted, I'd like to confirm that putting a nonsense name in the last position fails.  I'd also like to test that if your percentile has e.g. 99 and you ask for 95, that fails correctly.  I think it will, but let's make sure we have a test.

Also, just as a note, if instead of "addresses issue #..." you say "resolves #...", github will automatically close the issue when the PR is merged.

salvatore-campagna commented on server/src/main/java/org/elasticsearch/search/aggregations/metrics/AbstractInternalHDRPercentiles.java:
```diff
@@ -78,7 +78,14 @@ protected void doWriteTo(StreamOutput out) throws IOException {
     public double value(String name) {
-        return value(Double.parseDouble(name));
+        try {
+            return value(Double.parseDouble(name));
```
Correct...I didn't realise we can just compare the name (parameter) to the aggregation name 🤦

salvatore-campagna committed:
fix: just compare the last path entry (name) with the aggregation name

When the last path entry used as 'name' parameter matches the aggregation
name it means we are using a short path not including the metric name.
If those match we can just check if we just have one value to return
and use that.

salvatore-campagna committed:
Update docs/changelog/87220.yaml

salvatore-campagna committed:
test: add a test targeting an unknown aggregation

salvatore-campagna committed:
fix: do not rely on default set of percents

salvatore-campagna commented:
@elasticmachine update branch

salvatore-campagna commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into fix/72983-buckets-path-multi-value-agg

salvatore-campagna committed:
test: include some java tests

salvatore-campagna committed:
fix: rework percentiles bucket too

not-napoleon reviewed (approved):
LGTM, thanks for addressing the changes.
